### Analyse data Titanic Disaster

### Try Model: Tensor Flow Neural Network

In [1]:
# setup absolute path to location of package Starts and config-file 
from inspect import getsourcefile
import os.path as path, sys
current_dir = path.dirname(path.abspath(getsourcefile(lambda:0)))
sys.path.insert(0, current_dir[:current_dir.rfind(path.sep)])

# import package Starts
from Starts.startmodskl import *
from Starts.startmodtf import *

local_kwargs {'data_path': './data/train.csv, ./data/titanic_no_nan.csv ,./data/test.csv', 'drop_obj_col': False, 'nan_drop_col': False, 'nan_drop_row': False, 'nan_zero': False, 'nan_mean': True, 'nan_mean_neighbors': False}


Using TensorFlow backend.


In [2]:
nonan_data = data[1]
nonan_data.head()

Survived  Pclass  Sex        Age  Ticket    Fare  Title  FamilySize
0         0       3    1  29.699118     182   7.225      5           0
1         0       3    1  25.000000     328   7.775      5           1
2         0       3    1  24.000000     642   7.050      5           0
3         0       3    1  28.000000     542  22.525      5           0
4         0       3    1  29.699118     462  15.500      5           1

In [3]:
len(nonan_data.columns)

8

In [4]:
stf = StartModTFANN(nonan_data, label='Survived')
stf.info_help()

{'StartModTF.(data)': '\n        tbd\n\n        References:\n            https://github.com/tensorflow/models/tree/master/samples/cookbook/regression\n\n        :param data: Pandas-DataFrame\n        :return:\n        ',
 'info_help_StartModTF': 'StartModTF'}

In [5]:
stf.info_parameters()


Hidden_units: [10, 10] 

Optimizer: Adagrad 

Activation_function: relu 

Learning_Rate: 0.001 

Training_Steps: 1000 

Batch_Size: 10 

Number_of_epochs: 1 

Feature_Scaling: False 

Loss_function: mean_squared_error 



In [10]:
# Setup neural network parameters
input_dim = len(nonan_data.columns)
new_param={'hidden_units':[input_dim,2,1], 'optimizer':'Adam', 'activation_fn':'sigmoid', 'learning_rate': 0.0001,
           'steps':2000, 'batch_size':10, 'num_epochs':2, 'feature_scl':True, 'loss_fn':'binary_crossentropy'}
stf.update_parameters=new_param
stf.info_parameters()


Hidden_units: [8, 2, 1] 

Optimizer: Adam 

Activation_function: sigmoid 

Learning_Rate: 0.0001 

Training_Steps: 2000 

Batch_Size: 10 

Number_of_epochs: 2 

Feature_Scaling: True 

Loss_function: binary_crossentropy 



In [11]:
# Try method Keras Tensorflow to predict the Survivors (column 'Survived')
model, y_eval, y_pred = stf.keras_sequential()

Epoch 1/2
712/712 [==============================] - 0s 467us/step - loss: 0.6898 - acc: 0.6096
Epoch 2/2
712/712 [==============================] - 0s 74us/step

acc: 60.96%


In [12]:
# Convert predicted value by rounding it (>0.5 will be considered as 1 else 0)
y_pred = np.array([1 if item>0.5 else 0 for item in y_pred])

# Compare the predicted value with true_value
len(y_eval), len(y_pred)

(179, 179)

In [13]:
confusion_matrix(y_eval, y_pred)

array([[115,   0],
       [ 64,   0]])

### Result: 
#### the correct predicted result 115
#### the wrong predicted result 64

In [15]:
# update parameters
new_param={'hidden_units':[input_dim,10,1], 'optimizer':'Adam', 'activation_fn':'sigmoid', 'learning_rate': 0.0001,
           'steps':2000, 'batch_size':10, 'num_epochs':10, 'feature_scl':True, 'loss_fn':'binary_crossentropy'}
stf.update_parameters=new_param
model, y_true, y_pred = stf.keras_sequential()

Epoch 1/10
712/712 [==============================] - 0s 526us/step - loss: 0.6845 - acc: 0.6053
Epoch 2/10
712/712 [==============================] - 0s 203us/step - loss: 0.6623 - acc: 0.6096
Epoch 3/10
712/712 [==============================] - 0s 209us/step - loss: 0.6412 - acc: 0.6433
Epoch 4/10
712/712 [==============================] - 0s 282us/step - loss: 0.6232 - acc: 0.6896
Epoch 5/10
712/712 [==============================] - 0s 217us/step - loss: 0.6100 - acc: 0.6938
Epoch 6/10
712/712 [==============================] - 0s 208us/step - loss: 0.5926 - acc: 0.7022
Epoch 7/10
712/712 [==============================] - 0s 210us/step - loss: 0.5645 - acc: 0.7065
Epoch 8/10
712/712 [==============================] - 0s 231us/step - loss: 0.5233 - acc: 0.7598
Epoch 9/10
712/712 [==============================] - 0s 214us/step - loss: 0.4537 - acc: 0.8034
Epoch 10/10
712/712 [==============================] - 0s 188us/step

acc: 95.93%


In [16]:
# Convert predicted value by rounding it (>0.5 will be considered as 1 else 0)
y_pred = np.array([1 if item>0.5 else 0 for item in y_pred])
confusion_matrix(y_true, y_pred)

array([[112,   3],
       [  6,  58]])

### Result: 
#### the correct predicted result 170
#### the wrong predicted result 9 
#### the accuracy has been improved from 60.96% to 95.93%

#### Notice: unnecessary feature (e.g. 'Ticket') might be removed before building model to improve the result overall